src/profilers.cpp

```diff
@@ -2,19 +2,369 @@
+
+    for (int i = 0; i < parts.size(); i++) {
+      parts_c_str[i] = (char *)parts[i].c_str();
+    }
+    parts_c_str[parts.size()] = NULL;
+
+
+    pid_t forked_metric_profiler = fork(); //separate process for profiler, parent handles errors
+
+    if(forked_metric_profiler == 0){
+      int code_metric_exec = 0;
+
+      int status;
+      pid_t command_status = waitpid(pid, &status, WNOHANG); //check if profiled command has finished executing
+      const int interval_ns = 1000 * 1000000 / this->freq;
+
+      while (command_status > 0){
+
+        struct timespec start, end;
+        clock_gettime(CLOCK_MONOTONIC, &start);
+
+        int pipe_fd[2];
+
+        if (pipe(pipe_fd) == -1) {
+          print("Could not establish communication pipe between metric-exec "
+                "and metric-read in \"" + this->get_name() + "\"! Terminating "
+                "the profiled command wrapper.", true, true);
+          kill(pid, SIGTERM);
+          return;
+        }
+
+        pid_t forked_metric_exec = fork();
+        if (forked_metric_exec == 0){
+          close(pipe_fd[0]);
+
+          fs::current_path(result_processed);
+
+          int stderr_fd = creat(stderr_metric_command.c_str(), O_WRONLY);
+
+          if (stderr_fd == -1) {
+            std::exit(ERROR_STDERR);
+          }
+
+          if (dup2(stderr_fd, STDERR_FILENO) == -1) {
+            std::exit(ERROR_STDERR_DUP2);
+          }
+
+          if (dup2(pipe_fd[1], STDOUT_FILENO) == -1) {
+            std::exit(ERROR_STDOUT_DUP2);
+          }
+
+          close(pipe_fd[1]);
+          execvp(path, parts_c_str);
+
+          std::exit(errno);
+        }
+        else
+        {
+          close(pipe_fd[1]);
+          //read
+          char buffer[1024];
+          ssize_t bytes_read;
+          std::string data;
+          
+          while ((bytes_read = read(pipe_fd[0], buffer, sizeof(buffer) - 1)) > 0) {
+              buffer[bytes_read] = '\0';
+              data += buffer;
+          }
+          std::string parsed_data;
+
+          if(!this->regex.empty()){
+
+            std::regex number_regex(this->regex);
+
+            std::sregex_iterator numbers_begin = std::sregex_iterator(data.begin(), data.end(), number_regex);
+            std::sregex_iterator numbers_end = std::sregex_iterator();
+            for (std::sregex_iterator i = numbers_begin; i != numbers_end; ++i) {
+              std::smatch match = *i;
+
+              parsed_data += std::stoi(match.str());
+            }
+          }else{
+            parsed_data = data;
+          }
+
+          std::regex number_regex(R"(-?\d+(\.\d+)?)");;
+
+          std::smatch match;
+          std::string temp_data = parsed_data;
+          int count = 0;
+          float metric_val;
+                    
+          while (std::regex_search(temp_data, match, number_regex)) {
+            count++;
+            if(count > 1){break;}
+  
+            std::string number_str = match.str(0);
```

Comment:
The argument to ```match.str()``` may need to be 1 rather than 0, we'll see later when integrating this into the AdaptivePerf flow.

Reply:
Modified , although I am still unsure why it should be 1.